File: 21. Merge Two Sorted Lists/merge-two-sorted-lists.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        l1 = list1
        l2 = list2
        dummy = ListNode()
        current = dummy
        
        while l1 and l2:
            if l1.val < l2.val:
                current.next = l1
                l1 = l1.next
            else:
                current.next = l2
                l2 = l2.next
            current = current.next
            
        if l1:
            current.next = l1
        if l2:
            current.next = l2
            
        return dummy.next


list1 = ListNode(1, ListNode(2, ListNode(4)))
list2 = ListNode(1, ListNode(3, ListNode(4)))

s = Solution()

merge_list = s.mergeTwoLists(list1, list2)
result = []
current = merge_list
while current:
    result.append(current.val)
    current = current.next
print(result)

File: 21. Merge Two Sorted Lists/MergeTwoSortedLists.java
/**
 * Definition for singly-linked list. */
public class MergeTwoSortedLists {
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public class Solution {
        public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
            ListNode l1 = list1;
            ListNode l2 = list2;
            ListNode dummy = new ListNode(-1);
            ListNode current = dummy;

            while (l1 != null && l2 != null){
                if (l1.val < l2.val){
                    current.next = l1;
                    l1 = l1.next;
                } else {
                    current.next = l2;
                    l2 = l2.next;
                }
                current = current.next;
            }
            if (l1 != null){
                current.next = l1;
            }
            if (l2 != null){
                current.next = l2;
            }
            return dummy.next;
        }

        public void printList(ListNode head){
            ListNode current = head;
            while (current != null){
                System.out.print(current.val + " -> ");
                current = current.next;
            }
        }
    }

    public static void main(String[] args) {
        MergeTwoSortedLists outerInstance = new MergeTwoSortedLists();
        Solution solution = outerInstance.new Solution();

        ListNode list1 = outerInstance.new ListNode(1);
        list1.next = outerInstance.new ListNode(2);
        list1.next.next = outerInstance.new ListNode(4);

        ListNode list2 = outerInstance.new ListNode(1);
        list2.next = outerInstance.new ListNode(3);
        list2.next.next = outerInstance.new ListNode(4);

        ListNode mergedList = solution.mergeTwoLists(list1, list2);
        solution.printList(mergedList);
    }
}
